src/api/useViewportFeatures.js

```diff
@@ -71,8 +71,9 @@ export default function useViewportFeatures(source, uniqueId, debounceTimeOut =
   const viewport = useSelector((state) => state.carto.viewport);
   const [uniqueFeatures, setUniqueFeatures] = useState();
   const widgetsLoadingState = useSelector((state) => state.carto.widgetsLoadingState);
+  const referencedWidgetsLoadingState = useRef(widgetsLoadingState);
```

Comment:
Why using a ref here for this info from store?

Reply:
Because adding `widgetsLoadingState` in the dependency array [here](https://github.com/CartoDB/carto-react-lib/blob/27d5215aec0fcc0f86642b41fc1fea83cb3bf187/src/api/useViewportFeatures.js#L99) cause unnecessary/loop re-renders due to dispatching `setAllWidgetsLoadingState`. So, if state this store state is updated, `useEffect` will be called.
Wrapping things with `useRef` allows to persist the value for the entire lifetime of the component.